### PR TITLE
feat(storage-browser): add managed and test amplify auth support, refactor state layer and views

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
- add managed auth support and initial Amplify auth testing support
  - add `region`, `credentialsProvider` to `CreateStorageBrowserInput`
  - add `config` to `listLocationItemsAction`
  - add `createAmplifyAuthAdapter`
  - integrate `createManagedAuthConfigAdapter` with example app
  - integrate `createAmplifyAuthAdapter` with example app
  - add `bucket` to `LocationData` interface
- add `Permission` exclusion to `listLocationsAction`
- move `createListLocationsAction` handling to `createProvider`
- add `Controller` component for handling default behavior locations and location items loading
- seperate `LocationDetail` to `LocationDetailView` and `LocationActionView`
- remove `Provider`/`ViewTypeProvider` from `ViewComponent` for all Views
- add `location` selection to `LocationsView`
- integrate `LocationsDataProvider` with `ActionProvider`
- update `createActionStateContext` types

#### Issue #, if available
NA

#### Description of how you validated changes
Extensive manual testing
#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/Controller.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  LocationsDataState,
+  useAction,
+  useLocationsData,
+} from './context/actions';
+
+import { useControl } from './context/controls';
+import { useConfig } from './context/config';
+
+// will only work amplify buckets with access level set to "public"
+const TEMP_PREFIX = 'public/';
```
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/Controller.tsx]: Resolved, was a test variable

File: packages/react-storage/src/components/StorageBrowser/context/controls/Navigate/Navigate.tsx [hunk 2]
```diff
@@ -74,15 +43,33 @@ export function navigateReducer(
...
+//       if (initial === _next) {
+//         return prevState;
+//       }
+//       return { next: _next, previous };
+//     }
+//     default:
+//       throw new Error(`Invalid value of ${type} provided as \`type\``);
+//   }
+// }
+
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/context/controls/Navigate/Navigate.tsx:72]:
```suggestion
```

File: packages/react-storage/src/components/StorageBrowser/createStorageBrowser.tsx [hunk 3]
```diff
@@ -23,13 +22,20 @@ interface ResolvedStorageBrowserElements<
   T extends Partial<StorageBrowserElements>,
 > extends MergeBaseElements<StorageBrowserElements, T> {}
 
+/**
+ * Handles default `StorageBrowser` behavior:
+ * - render `LocationsView` on init
+ * - render `LocationDetailView` on location selection
+ * - render `ActionView` on action selection
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/createStorageBrowser.tsx:29]:
```suggestion
 * - TODO: render `ActionView` on action selection
```

File: packages/react-storage/src/components/StorageBrowser/Controller.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    type: 'LIST_LOCATION_ITEMS',
+  });
+  const [{ location }] = useControl({ type: 'NAVIGATE' });
+  const { current } = location;
+
+  const loadLocations = shouldLoadLocations(locationsState);
+
+  React.useEffect(() => {
+    if (loadLocations) {
+      // update to exhaustive call
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/Controller.tsx:42]:
```suggestion
      // TODO: update to exhaustive call
````

File: packages/react-storage/src/components/StorageBrowser/Controller.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+          region,
+        },
+      }),
+    [getLocationCredentials, handleListLocationItems, region]
+  );
+
+  const loadLocationItems = shouldLoadLocationItems(locationItemsState);
+  React.useEffect(() => {
+    if (current && loadLocationItems) {
+      listLocationItems(current);
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/Controller.tsx:67]:
```suggestion
      // TODO: update to exhaustive call
      listLocationItems(current);
````

File: packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/Controls.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { StorageBrowserElements } from '../../context/elements';
+import { Controls } from '../Controls';
+import { CommonControl } from '../types';
+
+export interface LocationActionViewControls<T extends StorageBrowserElements>
+  extends Pick<Controls<T>, CommonControl | 'Summary'> {
+  (): React.JSX.Element;
+}
+
+// @ts-expect-error Needs Controls assignment
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/Controls.tsx:12]:
```suggestion
// @ts-expect-error TODO: add Controls assignment
```

File: packages/react-storage/src/components/StorageBrowser/Views/LocationDetailView/Controls.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  > {
+  (): React.JSX.Element;
+}
+
+const TEMP_ACTIONS = [
+  { name: 'Upload File', type: 'UPLOAD_FILE' },
+  { name: 'Create Folder', type: 'CREATE_FOLDER' },
+];
+
+// @ts-expect-error needs Controls assignments
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/Views/LocationDetailView/Controls.tsx:21]:
```suggestion
// @ts-expect-error TODO: add Controls assignment
```

**Review by hbuchel — APPROVED** [review]
LGTM based on our walkthrough (feel free to ping me after merge conflicts are fixed)

**Review by esauerbo — APPROVED** [review]

**Review by esauerbo — APPROVED** [review]

**Review by hbuchel — APPROVED** [review]

**Pull request merged**